contracts/utils/Strings.sol

```diff
@@ -115,4 +128,247 @@ library Strings {
+     *
+     * This function will still revert if the result does not fit in a `uint256`
+     */
+    function tryParseHex(
+        string memory input,
+        uint256 begin,
+        uint256 end
+    ) internal pure returns (bool success, uint256 value) {
+        bytes memory buffer = bytes(input);
+
+        // skip 0x prefix if present
+        bool hasPrefix = bytes2(unsafeReadBytesOffset(buffer, begin)) == 0x3078;
+        uint256 offset = hasPrefix.toUint() * 2;
+
+        uint256 result = 0;
+        for (uint256 i = begin + offset; i < end; ++i) {
+            uint8 chr = _tryParseChr(buffer[i]);
+            if (chr > 15) return (false, 0);
+            result *= 16;
+            result += chr;
+        }
+        return (true, result);
+    }
+
+    /**
+     * @dev Parse a hexadecimal string (with or without "0x" prefix), and returns the value as an `address`.
+     *
+     * This function will revert if:
+     * - the string is not formated as `(0x)?[0-9a-fA-F]{40}`
+     */
+    function parseAddress(string memory input) internal pure returns (address) {
+        return parseAddress(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseAddress} that parses a substring of `input` located between position `begin` (included) and
+     * `end` (excluded).
+     */
+    function parseAddress(string memory input, uint256 begin, uint256 end) internal pure returns (address) {
+        (bool success, address value) = tryParseAddress(input, begin, end);
+        if (!success) revert StringsInvalidAddressFormat();
+        return value;
+    }
+
+    /**
+     * @dev Variant of {parseAddress-string} that returns false if the parsing fails because input is not a properly
+     * formated address.
+     */
+    function tryParseAddress(string memory input) internal pure returns (bool success, address value) {
+        return tryParseAddress(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseAddress-string-uint256-uint256} that returns false if the parsing fails because input is not a properly
+     * formated address.
+     */
+    function tryParseAddress(
+        string memory input,
+        uint256 begin,
+        uint256 end
+    ) internal pure returns (bool success, address value) {
+        // check that input is the correct length
+        bool hasPrefix = bytes2(unsafeReadBytesOffset(bytes(input), begin)) == 0x3078;
+        uint256 expectedLength = 40 + hasPrefix.toUint() * 2;
+
+        if (end - begin == expectedLength) {
+            // length garantees that this does not overflow, and value2 is at most type(uint160).max
+            (bool s, uint256 v) = tryParseHex(input, begin, end);
+            return (s, address(uint160(v)));
+        } else {
+            return (false, address(0));
+        }
+    }
+
+    // TODO: documentation.
+    function unsafeReadBytesOffset(bytes memory buffer, uint256 offset) internal pure returns (bytes32 value) {
```

Comment:
Where should that go ? A Bytes library ?

Reply:
Note that the same function exists in RSA.sol

Reply:
I believe this function was present in other places but always private.

Reply:
> Yes. private, and marked as memory safe, because the private calls are all enforcing the safety.

Right... That was the issue. That using memory-unsafe functions disables some optimizations globally, so they were important to avoid.

Reply:
So do we want to have:
- multiple private (and memory safe) implementations
- one internal, none memory safe implementation 
- one internal, memory safe (checks memory bound) implementation
?

Reply:
Extra option is:

- one internal, that is not actually memory safe, but that we mark as `"memory-safe"`. And we document it should only be used with input that are known to be within the bounds

Reply:
I'm not sure that the last option is a safe recommendation. I'm too stupid regarding how compiler optimizations work so feel free to disregard, but I wonder if guaranteeing the safetiness with a private function will have a negative effect on the so-called "‘stack-to-memory’ mover"